### ISSUE TITLE
command-line error: language modes specified are incompatible C++/C++ 1027

### ISSUE THREAD

**Issue by codewitch-honey-crisis** [microsoft/vscode-cpptools#7836]
Bug type: General

**Describe the bug**
- Windows 10 64 bit w/ MiniGW and GCC
- VS Code Version: 1.58.2
- C/C++ Extension Version: 1.51
- Other extensions you installed (and if the issue persists after disabling them): makefile tools, cmake tools (yes it persists. I disabled them. Then I removed them)
- 
- A clear and concise description of what the bug is.
I lose intellisense. I have these silly spurious errors in my output. My productivity is now garbage.

Please fix this. It wasn't happening on this project until today. Now I get this error in every source file I open.

**To Reproduce**
Get MiniGW and setup GCC

1. Go to https://github.com/codewitch-honey-crisis/nozip_repro
2. Click on Download Zip. Unzip it somewhere
3. Open the project

**Expected behavior**
I expect my intellisense to work.
I expect errors to actually reflect real errors in my source code and not random things that don't actually exist with no line numbers to tell me where it even went wrong and nothing I can change to remove it.

**Screenshots**
https://uploads.disquscdn.com/images/ee21fc7eaa54fd07e5bfd18bd643d861eca61e890b0f0d82390baaf28972381b.jpg 

None of the things in that screen shot should exist

**Additional context**
This didn't happen yesterday. It didn't happen when I woke up this morning. It shouldn't be happening now.

**Comment by Colengms** [comment]
Hi @codewitch-honey-crisis .  

I am unable to reproduce this issue with a clone of the repro you provided.  Could you please set `"C_Cpp.loggingLevel": "Debug"` and provide the output of the C/C++ output channel leading up to the repro?  Also, could you run the "Log Diagnostics" command with the file open that repros the error, and provide that output?

I assume you mean MinGW.  There are many different forks of the MinGW project.  What version of MinGW are you using and where did you download it from?

Since your repro did not include a `c_cpp_properties.json`, how are you configuring the C/C++ Extension?  If MinGW is the only compiler installed on your system, we should be using it by default.  But, if you are not configuring `compilerPath` directly, it's possible we may be finding and choosing another compiler found on your system.  The log output should indicate which compiler it is attempting to be compatible with.

**Comment by codewitch-honey-crisis** [comment]
I honestly don't remember where I got this MiniGW as I tried several before stumbling on a distro that I didn't find horrible.

I don't know where to set "C_Cpp.loggingLevel": "Debug" - what command in the command palette do i use for that, and or, what file do I create to put that into? I've never debugged one of your extensions before. Until now I haven't had to, because things worked.

I don't have a c_cpp_properties.json - I don't use those unless I have to 

Here's my tasks.json and launch.json below - those are the only vscode files in this project
```
{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "cppbuild",
			"label": "C/C++: g++.exe build active file",
			"command": "C:\\Program Files\\mingw-w64\\x86_64-8.1.0-posix-seh-rt_v6-rev0\\mingw64\\bin\\g++.exe",
			"args": [
				"-g",
				"${file}",
				"-o",
				"${fileDirname}\\${fileBasenameNoExtension}.exe"
			],
			"options": {
				"cwd": "${fileDirname}"
			},
			"problemMatcher": [
				"$gcc"
			],
			"group": {
				"kind": "build",
				"isDefault": true
			},
			"detail": "compiler: \"C:\\Program Files\\mingw-w64\\x86_64-8.1.0-posix-seh-rt_v6-rev0\\mingw64\\bin\\g++.exe\""
		}
	]
}
```

```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "g++.exe - Build and debug active file",
            "type": "cppdbg",
            "request": "launch",
            "program": "${fileDirname}\\${fileBasenameNoExtension}.exe",
            "args": [],
            "stopAtEntry": false,
            "cwd": "${fileDirname}",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "miDebuggerPath": "C:\\Program Files\\mingw-w64\\x86_64-8.1.0-posix-seh-rt_v6-rev0\\mingw64\\bin\\gdb.exe",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ],
            "preLaunchTask": "C/C++: g++.exe build active file"
        }
    ]
}
```

**Comment by Colengms** [comment]
> I don't know where to set `"C_Cpp.loggingLevel": "Debug"`

@codewitch-honey-crisis This would be specified in your `settings.json` file.  You can open that by running the command: `Preferences: Open Settings (JSON)`

Since you're not using a `c_cpp_properties.json` file, the extension would be using an auto-detected default configuration.  The output of the C/C++ output channel with debug logging enabled, and the output of the `Log Diagnostics` command, could provide us details on that detected configuration.

**Comment by codewitch-honey-crisis** [comment]
Thanks! My settings.json was created and now looks like this:```{    "C_Cpp.loggingLevel": "Debug",    "files.associations": {        "stdio.h": "c"    }}``` When I run the Log Diagnostics command I get this. I hope that’s what you were looking for. Looking at it, it looks like the first compilerPath attribute is wrong but I’m not sure why it would have that in there. I’ll look through my other config files again. ```-------- Diagnostics - 7/16/2021, 5:53:01 PMVersion: 1.5.1Current Configuration:{    "name": "Win32",    "includePath": [        "${workspaceFolder}/**"    ],    "defines": [        "_DEBUG",        "UNICODE",        "_UNICODE"    ],    "windowsSdkVersion": "10.0.19041.0",    "compilerPath": "C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.29.30037/bin/Hostx64/x64/cl.exe",    "cStandard": "c17",    "cppStandard": "c++17",    "intelliSenseMode": "windows-msvc-x64",    "compilerArgs": [],    "intelliSenseModeIsExplicit": false,    "cStandardIsExplicit": false,    "cppStandardIsExplicit": false,    "compilerPathIsExplicit": true,    "configurationProvider": "ms-vscode.cmake-tools",    "browse": {        "path": [            "${workspaceFolder}/**"        ],        "limitSymbolsToIncludedHeaders": true    }}Custom browse configuration: {    "browsePath": [        "d:/users/gazto/documents/nozip/src",        "d:/users/gazto/documents/nozip/build/cmakefiles/nozip.dir/src"    ],    "standard": "gnu99",    "compilerPath": "c:/program files/mingw-w64/x86_64-8.1.0-posix-seh-rt_v6-rev0/mingw64/bin/x86_64-w64-mingw32-gcc.exe",    "compilerArgs": [        "-g",        "-Wall",        "-Wextra",        "-Wno-unused-parameter",        "-std=gnu99"    ]}Translation Unit Mappings:[ D:\Users\gazto\Documents\nozip\nozip.h ]:    D:\USERS\GAZTO\DOCUMENTS\NOZIP\NOZIP.H[ D:\Users\gazto\Documents\nozip\src\unzip.cpp ]:    D:\USERS\GAZTO\DOCUMENTS\NOZIP\SRC\UNZIP.CPP[ D:\Users\gazto\Documents\nozip\src\zip.hpp ]:    D:\USERS\GAZTO\DOCUMENTS\NOZIP\SRC\ZIP.HPP[ D:\Users\gazto\Documents\nozip\src\stb_inflate.h ]:    D:\USERS\GAZTO\DOCUMENTS\NOZIP\SRC\STB_INFLATE.HTranslation Unit Configurations:[ D:\Users\gazto\Documents\nozip\nozip.h ]:    Process ID: 4820    Memory Usage: 10 MB    Compiler Path: c:/program files/mingw-w64/x86_64-8.1.0-posix-seh-rt_v6-rev0/mingw64/bin/x86_64-w64-mingw32-gcc.exe    Includes:        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\LIB\GCC\X86_64-W64-MINGW32\8.1.0\INCLUDE        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\LIB\GCC\X86_64-W64-MINGW32\8.1.0\INCLUDE-FIXED        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\X86_64-W64-MINGW32\INCLUDE    Standard Version: c++17    IntelliSense Mode: windows-gcc-x64    Other Flags:        --gcc        --gnu_version=80100        --header_only_fallback[ D:\Users\gazto\Documents\nozip\src\unzip.cpp ]:    Process ID: 2936    Memory Usage: 10 MB    Compiler Path: c:/program files/mingw-w64/x86_64-8.1.0-posix-seh-rt_v6-rev0/mingw64/bin/x86_64-w64-mingw32-gcc.exe    Includes:        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\LIB\GCC\X86_64-W64-MINGW32\8.1.0\INCLUDE        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\LIB\GCC\X86_64-W64-MINGW32\8.1.0\INCLUDE-FIXED        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\X86_64-W64-MINGW32\INCLUDE    Standard Version: c++17    IntelliSense Mode: windows-gcc-x64    Other Flags:        --gcc        --gnu_version=80100[ D:\Users\gazto\Documents\nozip\src\zip.hpp ]:    Process ID: 4296    Memory Usage: 10 MB    Compiler Path: c:/program files/mingw-w64/x86_64-8.1.0-posix-seh-rt_v6-rev0/mingw64/bin/x86_64-w64-mingw32-gcc.exe    Includes:        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\LIB\GCC\X86_64-W64-MINGW32\8.1.0\INCLUDE        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\LIB\GCC\X86_64-W64-MINGW32\8.1.0\INCLUDE-FIXED        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\X86_64-W64-MINGW32\INCLUDE    Standard Version: c++17    IntelliSense Mode: windows-gcc-x64    Other Flags:        --gcc        --gnu_version=80100        --header_only_fallback[ D:\Users\gazto\Documents\nozip\src\stb_inflate.h ]:    Process ID: 11556    Memory Usage: 10 MB    Compiler Path: c:/program files/mingw-w64/x86_64-8.1.0-posix-seh-rt_v6-rev0/mingw64/bin/x86_64-w64-mingw32-gcc.exe    Includes:        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\LIB\GCC\X86_64-W64-MINGW32\8.1.0\INCLUDE        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\LIB\GCC\X86_64-W64-MINGW32\8.1.0\INCLUDE-FIXED        C:\PROGRAM FILES\MINGW-W64\X86_64-8.1.0-POSIX-SEH-RT_V6-REV0\MINGW64\X86_64-W64-MINGW32\INCLUDE    Standard Version: c++17    IntelliSense Mode: windows-gcc-x64    Other Flags:        --gcc        --gnu_version=80100        --header_only_fallbackTotal Memory Usage: 40 MB ------- Workspace parsing diagnostics -------Number of files discovered (not excluded): 2386```  Sent from Mail for Windows 10 From: Colen Garoutte-CarsonSent: Friday, July 16, 2021 5:49 PMTo: microsoft/vscode-cpptoolsCc: honey the codewitch; MentionSubject: Re: [microsoft/vscode-cpptools] command-line error: language modes specified are incompatible C++/C++ 1027 (#7836) I don't know where to set "C_Cpp.loggingLevel": ***@***.*** This would be specified in your settings.json file. You can open that by running the command: Preferences: Open Settings (JSON)Since you're not using a c_cpp_properties.json file, the extension would be using an auto-detected default configuration. The output of the C/C++ output channel with debug logging enabled, and the output of the Log Diagnostics command, could provide us details on that detected configuration.—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by codewitch-honey-crisis** [comment]
Hi all After doing what you said and looking at the debug output, it appears VS Code was using a ghost (or default?) c_cpp_properties.json configuration set up for MSVC. I went down to configurations – the blue bar on the right, and noticed it said Win32 for some reason. I clicked on that, and it gave me the option to create one which made a new json file. I then was able to edit the file using the UI to set it to GCC and I got everything working. Thanks. Sent from Mail for Windows 10 From: Colen Garoutte-CarsonSent: Friday, July 16, 2021 5:49 PMTo: microsoft/vscode-cpptoolsCc: honey the codewitch; MentionSubject: Re: [microsoft/vscode-cpptools] command-line error: language modes specified are incompatible C++/C++ 1027 (#7836) I don't know where to set "C_Cpp.loggingLevel": ***@***.*** This would be specified in your settings.json file. You can open that by running the command: Preferences: Open Settings (JSON)Since you're not using a c_cpp_properties.json file, the extension would be using an auto-detected default configuration. The output of the C/C++ output channel with debug logging enabled, and the output of the Log Diagnostics command, could provide us details on that detected configuration.—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by Colengms** [comment]
Hi @codewitch-honey-crisis  FYI, when you reply via email, your messages don't seem to get interpreted correctly by github.  I've edited them on github to restore the newlines so I better understand them.  :)

I'm glad this is working properly for you now.  The configuration information you saw that referred to "cl.exe" was actually the default configuration that was detected.

Do you have a `CMakeFiles.txt` file in your folder that was not included in the repro repo?  In your log diagnostics output, it appears that CMake Tools registered with the C.C++ extension (which it should only be doing if a `CMakeFiles.txt` is present).  As a result, we were using configuration information provided by CMake Tools instead, which was set to use MinGW.  If you wish to use CMake Tools, you could add the following to your `c_cpp_properties.json`: `"configurationProvider": "ms-vscode.cmake-tools"`.  If that makes your original repro come back, there may still be an issue here for us to fix.

**Comment by codewitch-honey-crisis** [comment]
Thanks for editing that - I'll reply from Github directly from here out. I had CMake installed, and it created a CMake for me I think? I could be wrong as I originally downloaded some code from github to fiddle with and learn from. I don't think it had a cmake but i know i ended up with one, which I deleted. At the time, i also had the cmake and make extensions installed. I have since removed them both. I hope to reinstall them later, but the make one particularly is an early edition and it leaves me feeling antsy.
